cmake/kokkos_arch.cmake

```diff
@@ -389,7 +396,35 @@ IF (KOKKOS_ENABLE_OPENMPTARGET)
 ENDIF()
 
 IF(KOKKOS_ENABLE_CUDA AND NOT CUDA_ARCH_ALREADY_SPECIFIED)
-  MESSAGE(SEND_ERROR "CUDA enabled but no NVIDIA GPU architecture currently enabled.  Please give one -DKokkos_ARCH_{..}=ON' to enable an NVIDIA GPU architecture.")
+  # Try to autodetect the CUDA Compute Capability by asking the device
+  SET(_binary_test_dir ${CMAKE_CURRENT_BINARY_DIR}/cmake/compile_tests/CUDAComputeCapabilityWorkdir)
+  FILE(REMOVE_RECURSE ${_binary_test_dir})
+  FILE(MAKE_DIRECTORY ${_binary_test_dir})
+
+  EXECUTE_PROCESS(
+    COMMAND ${CMAKE_CXX_COMPILER}
+      ${CMAKE_CURRENT_SOURCE_DIR}/cmake/compile_tests/cuda_compute_capability.cu
+      -o cuda_compute_capability
+    WORKING_DIRECTORY ${_binary_test_dir}
+    OUTPUT_QUIET
+    ERROR_QUIET)
+  EXECUTE_PROCESS(
+    COMMAND ${_binary_test_dir}/cuda_compute_capability
+    RESULT_VARIABLE _result
+    OUTPUT_VARIABLE CUDA_COMPUTE_CAPABILITY)
+  IF(${_result} EQUAL 0)
+    LIST(FIND KOKKOS_CUDA_ARCH_FLAGS sm_${CUDA_COMPUTE_CAPABILITY} FLAG_INDEX)
+    IF(FLAG_INDEX EQUAL -1)
+      MESSAGE(SEND_ERROR "Unsupported compute capability ${CUDA_COMPUTE_CAPABILITY}.")
+    ELSE()
+      MESSAGE(STATUS "Detected CUDA Compute Capability ${CUDA_COMPUTE_CAPABILITY}")
+      LIST(GET KOKKOS_CUDA_ARCH_LIST ${FLAG_INDEX} ARCHITECTURE)
+      KOKKOS_SET_OPTION(ARCH_${ARCHITECTURE} ON)
+      CHECK_CUDA_ARCH(${ARCHITECTURE} sm_${CUDA_COMPUTE_CAPABILITY})
+    ENDIF()
+  ELSE()
+    MESSAGE(SEND_ERROR "CUDA enabled but no NVIDIA GPU architecture currently enabled and auto-detection failed. Please give one -DKokkos_ARCH_{..}=ON' to enable an NVIDIA GPU architecture.")
```

Comment:
> Is there a reason we do this with two execute_process calls instead of try_run?

This is what I had already. I can have a look to use `try_run` instead if you prefer.

> The most likely reason auto-detection fails is compiling on a login node. Is there a good way to 1) tell them to try configuring from a compute node? or 2) make the utility easy to build and run?
 
We can mention that in the error message. In the end, the source file should be stand-alone anyway.